build.sh

```diff
@@ -1,65 +1,110 @@
+  echo "Usage: $0 <godot branch> <mono version> [<mono branch> <mono commit hash>]"
   echo
   echo "Examples: $0 3.1 mono-5.18.1.3"
-  echo "          $0 master mono-6.6.0.160 2019-08 bef1e6335812d32f8eab648c0228fc624b9f8357"
+  echo "	$0 master mono-6.6.0.160 2019-08 bef1e6335812d32f8eab648c0228fc624b9f8357"
+  echo
+  echo "godot branch:"
+  echo "mono version:"
+  echo "	These are combined to form the docker image tag, e.g. 'master-mono-6.6.0.160'."
+  echo "	Git will then clone the branch/tag that matches the mono version."
+  echo 
+  echo "mono branch:"
+  echo "	If specified, git will clone this mono branch/tag instead. Requires specifying a commit."
   echo
+  echo "mono commit:"
+  echo "	If specified, git will check out this commit after cloning."
+  echo 
   exit 1
 fi
 
 godot_branch=$1
 mono_version=$2
 img_version=$godot_branch-$mono_version
+files_root=$(pwd)/files
 mono_commit=
+mono_commit_str=
+
+# If optional Mono git branch and commit hash were passed, use them.
 if [ ! -z "$3" -a ! -z "$4" ]; then
-  # Optional Mono git branch and commit hash were passed,
-  # use that for the git clones.
   mono_version=$3
   mono_commit=$4
+  mono_commit_str="-${mono_commit:0:6}"
+fi
+
+# If mono branch does not start with mono-, prepend it to the folder name.
+if [ ${mono_version:0:5} != "mono-" ]; then
+  mono_root="${files_root}/mono-${mono_version}${mono_commit_str}"
+else
+  mono_root="${files_root}/${mono_version}${mono_commit_str}"
+fi
+
+# Confirm settings
+echo "Docker image tag: ${img_version}"
+echo "Mono branch: ${mono_version}"
+if [ ! -z "$mono_commit" ]; then
+  echo "Mono commit: ${mono_commit}"
 fi
-echo "Building images with version: ${img_version}"
-echo "Mono version used: ${mono_version} ${mono_commit}"
+if [ -e ${mono_root} ]; then
+  mono_exists="(exists)"
+fi
+echo "Mono source folder: ${mono_root} ${mono_exists}"
 echo
 while true; do
-    read -p "Is this correct? [y/n] " yn
-    case $yn in
-        [Yy]* ) break;;
-        [Nn]* ) exit 1;;
-        * ) echo "Please answer yes or no.";;
-    esac
+  read -p "Is this correct? [y/n] " yn
+  case $yn in
+    [Yy]* ) break;;
+    [Nn]* ) exit 1;;
+    * ) echo "Please answer yes or no.";;
+  esac
 done
 
 mkdir -p logs
 
+# Check out and patch Mono version
+if [ ! -e ${mono_root} ]; then
+  if [ ! -z "${mono_commit}" ]; then 
+    # If a commit is specified, get the full history
+    git clone -b ${mono_version} --single-branch --progress https://github.com/mono/mono ${mono_root}
+    pushd ${mono_root}
+    git checkout ${mono_commit} 
+  else
+    # Otherwise, get a shallow repo
+    git clone -b ${mono_version} --single-branch --progress --depth 1 https://github.com/mono/mono ${mono_root}
+    pushd ${mono_root}
+  fi
+  # Download all submodules, up to 6 at a time
+  git submodule update --init --recursive --recommend-shallow -j 6 --progress
+  patch -p1 < ${files_root}/patches/mono-unity-Clear-TLS-instead-of-aborting.patch 
+  popd
+fi
+
```

Comment:
`-j 6` made a huge difference in total download time. Less timeout and waiting because it was downloading other modules while one was timing out.